### ISSUE TITLE
基础组件

### ISSUE THREAD

**Issue by NayukiChiba** [NayukiChiba/CNN#10]

**Comment by NayukiChiba** [comment]
src/model/ 模块需要写 3 个文件，按依赖顺序：

---

### 清单

| # | 文件 | 职责 | 核心内容 | 依赖 |
|---|---|---|---|---|
| 1 | `layers.py` | 基础组件 | `ConvBlock` — Conv2d + BatchNorm2d + ReLU + 可选 MaxPool | 无 |
| | | | `LinearBlock` — Linear + BatchNorm1d + ReLU + 可选 Dropout | |
| 2 | `cnn.py` | CNN 模型 | 用上面的 Block 组装网络：ConvBlock×2 → Flatten → LinearBlock×2 → Linear(10) | `layers.py` |
| | | | `forward()` 返回 logits | |
| 3 | `factory.py` | 模型工厂 | 根据参数字典构建 `CNN` 实例，自动 `.to(device)` | `cnn.py` |

---

### 各文件拆开看

**`layers.py`** — 两个 Block 类：

```python
class ConvBlock:
    输入 → Conv2d(inC, outC, kernel) → BatchNorm2d(outC) → ReLU → [MaxPool(2)] → 输出
    参数: in_channels, out_channels, kernel_size, pool (bool)

class LinearBlock:
    输入 → Linear(inF, outF) → BatchNorm1d(outF) → ReLU → [Dropout(p)] → 输出
    参数: in_features, out_features, dropout (float or None)
```

**`cnn.py`** — 组装网络：

```
Input (1, 28, 28)
  → ConvBlock(1, 32, pool=True)   → (32, 14, 14)
  → ConvBlock(32, 64, pool=True)  → (64, 7, 7)
  → Flatten                       → (64*7*7)
  → LinearBlock(3136, 128, dp=0.5) → (128)
  → Linear(128, 10)               → (10)  ← logits
```

**`factory.py`** — 一行创建模型：

```python
model = createModel(modelParams, device="cuda")
# 等价于 model = CNN(...).to("cuda")
```

---

**Comment by NayukiChiba** [comment]
---

## 目标是什么

不是"设计最好的网络"，是**让模型学会从像素里认出数字**。翻译成工程问题：

```
输入：一张 28×28 的灰度图（784 个数）
输出：0-9 之间的一个数字

本质上就是找一个函数 f(784个数) → 1个数
```

为什么不能直接 `y = Wx + b`（784 → 10）？

因为**像素位置是死的，数字位置不是死的**。同一张图里，数字可能偏左、偏右、写大、写小。784 维的线性模型没有平移不变性——位置一变，它就不认识了。

所以目标很具体：**找一个函数，不管数字写在图片哪个位置、大小略有变化、粗细不同，都能认出来。**

CNN 怎么满足这个目标？

| CNN 的操作 | 解决的问题 |
|---|---|
| 卷积（小窗口滑动） | 平移不变性：同一笔画出现在图片不同位置，同一个卷积核都能检测到 |
| 池化（压缩分辨率） | 大小不变性 + 减少位置敏感度 |
| 多层堆叠 | 从简单特征（边缘）组合出复杂特征（形状） |
| 全连接最后一层 | 把空间特征汇总成分类决策 |

---

## 这个结构是怎么"想"出来的

不是凭空想，是**三十年渐进试出来的**。

### 1990 之前：模板匹配

```
存一张标准"7"的图片，新图来了逐像素对比，相似度高就是 7
```
失败原因：没有人写出来的 7 长一样。

### 1990：LeNet（Yann LeCun）

第一次有人把"特征提取"和"分类"分开：

```
输入图片
  → 手工写几个检测器（检测边缘/拐角）  ← 手动设计
  → 把这些检测结果汇总
  → 分类
```

LeCun 发现：为什么手工设计检测器？让网络自己学不就行了。于是第一个 CNN 诞生了——LeNet-5，结构跟你现在要写的一模一样：

```
Conv → Pool → Conv → Pool → FC → FC → softmax
```

这篇 1998 年的论文用 MNIST 手写数字做了验证，就是你这个项目的鼻祖。

### 2012：AlexNet

LeNet 但更大、更深、用 GPU 训练。ImageNet 上碾压一切传统方法。从此大家疯狂堆层。

### 2014：VGG

实验结论：**用小卷积核（3×3）堆更多层，效果比大卷积核（7×7）好**。

两个 3×3 的感受野 = 一个 5×5，但参数少了一半，还非线性的次数更多。

### 2015：BatchNorm

训练深层网络时，前面层的参数一变，后面层的输入分布就变。BatchNorm 把每一层的输出强行拉回零均值单位方差，训练稳定了、学习率可以开更大、收敛更快。

### 2015 至今

核心套路凝固了：**Conv → BN → ReLU → Pool** 反复堆，最后接 **FC → BN → ReLU → Dropout**，然后分类头。这个配方是二十年的实验结果，不是理论推导。

---

## 中间的优化过程

训练一个 CNN 不是"跑一次就完"，而是一个调试循环：

```
初始化参数（随机）
    │
    ▼
前向传播：输入 → 网络 → 输出 logits
    │
    ▼
计算损失： CrossEntropyLoss(logits, 真实标签)
    │
    ▼
反向传播： 损失对每个参数求梯度
    │
    ▼
参数更新： weight = weight - lr × 梯度  (Adam 优化器)
    │
    ▼
重复 N 个 epoch，看验证集准确率变化
```

你实际会遇到的调试点：

| 现象 | 原因 | 调整方向 |
|---|---|---|
| 训练 loss 不降 | 学习率太小 / 网络太深梯度消失 | 加大 lr / 加 BN / 检查数据 |
| 训练 loss 降但 val 不降 | 过拟合 | 加 Dropout / 加 weight decay / 减模型 |
| loss 震荡不收敛 | 学习率太大 | 减小 lr |
| val loss 长期不降 | 学习率在平台期 | ReduceLROnPlateau（你 config 里已经配好了） |
| 训练极慢 | batch_size 太小 / num_workers 太少 | 调整 dataloader 参数 |

---

## 怎么理解这个过程

五个层次，逐级深入：

**第一层（会用）**：找一个现成 CNN 代码，改改参数（通道数、层数、lr），跑起来，看 loss 曲线。打 Label → 训练 → 预测，走通全流程。

**第二层（理解每个零件）**：把每一层拆出来，打印前后的 shape 变化。你的项目做这个很容易——forward 里加几行 print，跑一个 batch 就能看到。

**第三层（知道为什么）**：读 LeNet-5 和 AlexNet 论文。不长，各十几页。然后你就知道"这个结构是 1998 年就在用的"、"BN 是为了解决梯度消失"、"Dropout 是为了防过拟合"。

**第四层（能调参）**：固定一个 baseline，每次只改一个参数（通道数 / 层数 / lr / dropout），画 val accuracy 曲线对比。记住哪个参数影响什么。

**第五层（能自己搭）**：拿到一个新的图像分类任务，第一反应是"先跑个 ResNet-18，然后根据数据量决定要不要自己写小网络"。

你现在在第一层和第二层之间。**重点是先跑起来**。你把这个项目从头写完——dataset → dataloader → model → train → eval → inference——整条链路通了以后，回来看这段解释，就会有一种"原来如此"的感觉。
